g3doc/Tutorials/Toy/Ch-2.md

```diff
@@ -63,88 +78,89 @@ in MLIR the location is a core requirement which translates in APIs manipulating
 operations requiring it. Dropping a location becomes an explicit choice and
 cannot happen by mistake.
 
+## Opaque API
 
-## Opaque Builder API
-
-Operations and types can be created with only their string names using the
-raw builder API. This allows MLIR to parse, represent, and round-trip any valid
-IR. For example, the following can round-trip through *mlir-opt*:
+MLIR is designed to be a completely extensible system, as such the
+infrastructure has the capability to opaquely represent operations (as well as
+attributes, types, etc.) that have not been registered. This allows MLIR to
+parse, represent, and round-trip any valid IR. For example, the following can
+round-trip through *mlir-opt*:
 
 ```MLIR(.mlir)
 func @some_func(%arg0: !random_dialect<"custom_type">) -> !another_dialect<"other_type"> {
-  %result = "custom.operation"(%arg0) : (!random_dialect<"custom_type">) -> !another_dialect<"other_type">
+  %result = "custom.operation"(%arg0) { attr = #random_dialect<"custom_attribute"> } : (!random_dialect<"custom_type">) -> !another_dialect<"other_type">
   return %result : !another_dialect<"other_type">
 }
 ```
 
 Here MLIR will enforce some structural constraints (SSA, block termination,
-return operand type coherent with function return type, etc.) but otherwise the
-types and the operation are completely opaque.
+etc.) but otherwise the types and the `custom.operation` are completely opaque.
 
-We will take advantage of this facility to emit MLIR for Toy by traversing the
-AST. Our types will be prefixed with "!toy" and our operation name with "toy.".
-MLIR refers to this prefix as a *dialect*, we will introduce this with more
-details in the [next chapter](Ch-3.md).
+We will take advantage of this facility for the initial emission of MLIR for Toy
```

Comment:
I thought something you wanted was to include a more prominent warning about the fact that this is only for demo purpose, and pointing that the next chapter will introduce the more common way of creating IR? (or something like that).

Reply:
I also like the idea of introducing IR registration first but then mentioning that the IR will roundtrip without registration but won't do any checking, without this being a part of the actual tutorial source.

Reply:
I'm not sure of the exact flow of Ch.2 and Ch.3 at this point and they are likely to be merged. At this point I'm just updating/modernizing the chapter. I'm leaving the wording for this warning to a followup.